libtcmur/libtcmur.3
libtcmur(3)                 Linux Programmer's Manual                libtcmur(3)

NAME
    libtcmur -- usermode API to tcmu-runner block storage handlers

SYNOPSIS
    #include "libtcmur.h"

    error_t libtcmur_init(const char * handler_prefix);
    error_t libtcmur_exit(void);

    error_t tcmur_handler_load(const char * subtype);
    error_t tcmur_handler_unload(const char * subtype);

    error_t tcmur_check_config(char const * cfgstring);
    error_t tcmur_device_add(int minor, const char * cfgstring);
    error_t tcmur_device_remove(int minor);

    error_t tcmur_read(int minor, struct tcmulib_cmd *,
			struct iovec *, size_t niov, size_t, loff_t);
    error_t tcmur_write(int minor, struct tcmulib_cmd *,
			struct iovec *, size_t niov, size_t, loff_t);
    error_t tcmur_flush(int minor, struct tcmulib_cmd *);

    ssize_t tcmur_get_size(int minor);
    ssize_t tcmur_get_block_size(int minor);
    ssize_t tcmur_get_max_xfer(int minor);

    const char * tcmur_get_dev_name(int minor);

DESCRIPTION
    libtcmur provides a usermode application programming interface to access
    block storage services through tcmu-runner block storage handlers.

    Note that LIO, TCMU and tcmu-runner are uninvolved -- libtcmur only calls
    the tcmu-runner loadable storage handlers, e.g. qcow, glfs, ram, etc.

    libtcmur makes use of the handler read, write, and flush block I/O
    functions only -- no calls are made to handle_cmd().

    Functions returning type error_t return zero for success, otherwise -errno.

    Call libtcmur_init() once before using libtcmur services.  If handler_prefix
    is NULL, the default is used: "/usr/local/lib/tcmu-runner/handler_".  The
    expected handler paths are the concatenation of:

		    handler_prefix  tcmu_subtype  ".so"

    Call libtcmur_exit() once last, after any other functions described here.

    tcmur_handler_load() will dlopen() a tcmu-runner handler of the given
    subtype and load it into the program for use.  tcmur_handler_unload()
    unloads it.

    tcmur_check_config() checks a handler device configuration string for
    validity.  The handler for the configuration must already have been loaded
    using tcmur_handler_load().  cfgstring takes this form, specifying the
    handler's TCMU subtype:
		    /subtype/handler-cfg-string

    See tcmu-runner(8) for more about subtype and handler-cfg-string.

    tcmur_device_add() adds a device, with specified cfgstring, as the
    specified tcmur minor number.  tcmur_device_remove() removes the specified
    minor.  The handler is determined from the subtype in the first segment of
    cfgstring.  All tcmur subtypes share a common space of minor numbers.

    tcmur_read(), tcmur_write() and tcmur_flush() start I/O operations to the
    specified minor.  Errors in the I/O start process can be reported by -errno
    return from these calls.  A return value of zero denotes a successful I/O
    start, in which case there will be a completion call to cmd->done(), which
    may report either an "sts" error, or success (denoted by TCMU_STS_OK).

    Note that the completion call may occur before the request call returns.

    tcmur_read() and tcmur_write() take an iovec array with niov elements, an
    I/O size in bytes, and a seek offset into the device where the I/O begins.

    struct tcmulib_cmd includes this field, which must be set before passing
    the command to tcmur_read(), tcmur_write(), or tcmur_flush():

	cmd_done_t done;	/* completion callback */

    The callback function is of this type:

	typedef
	void (*cmd_done_t)(struct tcmu_device *, struct tcmulib_cmd *, int);

    The third argument to the callback is TCMU status (see libtcmu_common.h).

    tcmur_get_size() returns the size in bytes of the specified minor.  If the
    minor does not exist then the return is a -errno.

    tcmur_get_block_size() returns the block size in bytes of the specified
    minor.  If the minor does not exist then the return is a -errno.

    tcmur_get_max_xfer() returns the maximum I/O size in bytes of the specified
    minor.  If the minor does not exist then the return is a -errno.

    tcmur_get_dev_name() returns the device name of the specified minor.  If
    the minor does not exist then the return is NULL.

RETURN VALUE
    Upon successful completion, functions returning type error_t return zero.
    All functions return -errno on failure, except tcmur_get_dev_name(), which
    returns NULL in that case.

ERRORS
    tcmur_read()
    tcmur_write()
	-ENODEV	    no device at specified minor (including minor out of range)
	-ENXIO	    handler does not implement the requested function
	-EINVAL	    I/O would exceed device bounds
	-EIO	    I/O completed with nonzero "sts"

    tcmur_flush()
	-ENODEV	    no device at specified minor
	-EIO	    I/O completed with nonzero "sts"

    tcmur_get_size(), tcmur_get_block_size(), tcmur_get_max_xfer()
	-ENODEV	    no device at specified minor

    tcmur_device_add()
	-ENODEV	    minor number out of range
	-EBUSY	    minor number already in use by prior add
	errors returned by tcmur_check_config()
	errors returned by rhandler->open()

    tcmur_device_remove()
	-ENODEV	    no device at specified minor

    tcmur_check_config()
	-ENXIO	    no loaded handler subtype matches this config string
	-EINVAL	    config string does not start with '/'
	-EINVAL	    config string is too long
	errors returned by rhandler->check_config()

    tcmur_handler_load()
	-EEXIST	    handler already loaded for the specified subtype
	-ENOSPC	    all handler slots are in use
	-ENOMEM	    failed to asprintf the handler path
	-ENOENT	    failed to dlopen the handler path
	-EBADF	    failed to dlsym("handler_init")
	-EIO	    handler_init returned non-zero

    tcmur_handler_unload()
	-ENOENT	    no handler is loaded for the specified subtype
	-EBUSY	    handler has existing devices (added but not removed)

    libtcmur_exit()
	-EBUSY	    a handler is still loaded

NOTES

BUGS
    Not all symbols possibly referenced by handlers are implemented.
    Some such symbols are "stubbed out" and print a warning if called.
    Others do not exist, which will disallow loading of referencing handlers.

    There is no check for an attempt to add the same device twice.

    There is no generic way to specify the device size, block size, or maximum
    I/O size.

    libtcmur_exit() should auto-unload any handlers that have no devices
    currently added.

SEE ALSO
    tcmu-runner(8)

AUTHOR
    David A. Butterfield

Manpage updated 23 Jul 2019


libtcmur/libtcmur.c
/* libtcmur.c -- Usermode API to tcmu-runner block storage handlers
 *
 * Copyright (c) 2019 David A. Butterfield
 *
 * This file is licensed to you under your choice of the GNU Lesser
 * General Public License, version 2.1 or any later version (LGPLv2.1 or
 * later), or the Apache License 2.0.
 *
 * libtcmur provides a usermode application programming interface to access
 * block storage services through tcmu-runner block storage handlers.
 *
 * LIO, TCMU and tcmu-runner are not involved -- this only uses the tcmu-runner
 * loadable storage handlers, e.g. qcow, glfs, ram, etc.
 *
 * This library makes use of the handler read, write, and flush block I/O
 * functions only -- no calls are made to handle_cmd().
 */
#define _GNU_SOURCE
#include <sys/types.h>
#include <inttypes.h>
#include <stdbool.h>
#include <errno.h>
#include <stddef.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <dlfcn.h>

#include "libtcmur.h"
#include "sys_impl.h"

#define tcmu_io_trace_dev(dev, fmtargs...)  //  tcmu_dev_info(dev, "libtcmur: "fmtargs)

/* Print up to two stacktraces of callers per stubbed function */
#define STUB_WARN() do { \
    static int been_here = 0; \
    if (been_here++ < 2) \
	sys_backtrace("CALLED STUB FUNCTION %s\n", __func__); \
} while (0)

/* Handlers may have code that calls these functions, even though that code
 * would only ever be executed if we called handle_command(), which we don't.
 * But it is unexpected so we want to know if and whence they're ever called.
 */

/* SCSI-specific */
extern int tcmur_handle_caw(struct tcmu_device *, struct tcmur_cmd *, void * fn);
extern int tcmur_handle_writesame(struct tcmu_device *, struct tcmur_cmd *, void * fn);
extern void tcmu_notify_lock_lost(struct tcmu_device *);

uint32_t tcmu_dev_get_opt_unmap_gran(struct tcmu_device *dev) { STUB_WARN(); return 0; }
bool tcmu_dev_get_unmap_enabled(struct tcmu_device *dev) { STUB_WARN(); return false; }
uint32_t tcmu_dev_get_unmap_gran_align(struct tcmu_device *dev) { STUB_WARN(); return 0; }
void tcmu_dev_set_max_unmap_len(struct tcmu_device *dev, uint32_t len) { STUB_WARN(); }
void tcmu_dev_set_opt_unmap_gran(struct tcmu_device *dev, uint32_t len, bool split) { STUB_WARN(); }
void tcmu_dev_set_opt_xcopy_rw_len(struct tcmu_device *dev, uint32_t len) { STUB_WARN(); }
char * tcmu_cfgfs_dev_get_wwn(struct tcmu_device *dev) { STUB_WARN(); return NULL; }
int tcmur_handle_caw(struct tcmu_device *dev, struct tcmur_cmd *cmd, void * fn) { STUB_WARN(); return -1; }
int tcmur_handle_writesame(struct tcmu_device *dev, struct tcmur_cmd *cmd, void * fn) { STUB_WARN(); return -1; }
void tcmu_notify_lock_lost(struct tcmu_device *dev) { STUB_WARN(); }

/* Possibly applicable to libtcmur (XXX possible enhancement) */
extern void tcmu_notify_conn_lost(struct tcmu_device *);
extern int tcmur_dev_update_size(struct tcmu_device *, uint64_t new_size);

uint64_t tcmu_cfgfs_dev_get_info_u64(struct tcmu_device *dev, const char *name, int *fn_ret) { STUB_WARN(); return 0; }
int tcmu_make_absolute_logfile(char *path, const char *filename) { STUB_WARN(); return -1; }
bool tcmu_dev_get_solid_state_media(struct tcmu_device *dev) { STUB_WARN(); return false; }
void tcmu_dev_set_solid_state_media(struct tcmu_device *dev, bool solid_state) { STUB_WARN(); }
void tcmu_notify_conn_lost(struct tcmu_device *dev) { STUB_WARN(); }
int tcmur_dev_update_size(struct tcmu_device *dev, uint64_t new_size) { STUB_WARN(); return -1; }

/******************************************************************************/

const char * libtcmur_version = "libtcmur " TCMUR_VERSION;

static const char * handler_prefix = DEFAULT_HANDLER_PATH "/handler_";
static struct tcmur_handler * the_tcmu_handlers[MAX_TCMUR_HANDLERS];
static struct tcmu_device * the_tcmu_devices[MAX_TCMUR_MINORS];

#define device_of_minor(minor) \
  ((minor) < (int)ARRAY_SIZE(the_tcmu_devices) ? the_tcmu_devices[minor] : NULL)

struct tcmu_device {
	uint64_t num_lbas;
	uint32_t block_size;
	uint32_t max_xfer_len;
	unsigned int write_cache_enabled:1;
	unsigned int solid_state_media:1;
	char dev_name[16];		    /* e.g. "qcow3" */
	char cfgstring[PATH_MAX];
	char cfgstring_orig[PATH_MAX];
	void *hm_private;
	struct tcmur_handler *rhandler;
};

/* Find handler whose subtype string matches the subtype argument,
 * and pass back its slot number
 */
static struct tcmur_handler *
tcmur_find_handler_slot(const char * subtype, int *slotp)
{
	int i;
	for (i = 0; i < (int)ARRAY_SIZE(the_tcmu_handlers); i++) {
		struct tcmur_handler * h = the_tcmu_handlers[i];
		if (h)
		if (h && !strcmp(h->subtype, subtype)) {
			*slotp = i;
			return h;
		}
	}
	*slotp = -1;
	return NULL;
}

/* Find handler whose subtype string matches the subtype argument */
static struct tcmur_handler *
tcmur_find_handler(const char * subtype)
{
	int slot;
	return tcmur_find_handler_slot(subtype, &slot);
}

/* Return the handler that corresponds to cfgstr */
static struct tcmur_handler *
handler_of_cfgstr(const char * cfg)
{
    struct tcmur_handler * handler;
    char * hname;
    const char * subtype = cfg;
    const char * p;
    int ret;

    while (*subtype == '/')
	subtype++;
    p = subtype;
    while (isalnum(*p))
	p++;

    ret = sys_asprintf(&hname, "%.*s", (int)(p - subtype), subtype);
    verify_ge(ret, 0, "asprintf: %s", strerror(errno));

    handler = tcmur_find_handler(hname);
    sys_mem_free(hname);
    return handler;
}

/* Call a handler's check_config function */
error_t
tcmur_check_config(const char * cfg)
{
    error_t err;
    char * reason = NULL;
    struct tcmur_handler * handler;
    struct tcmu_device * dev = NULL;

    if (!cfg || *cfg != '/') {
	tcmu_err("config string must start with '/': '%s'\n", cfg);
	return -EINVAL;
    }

    if (strlen(cfg) >= sizeof(dev->cfgstring)) {
	tcmu_err("cfg string too long (%lu/%lu): '%s'\n",
		 strlen(cfg), sizeof(dev->cfgstring)-1, cfg);
	return -EINVAL;
    }

    handler = handler_of_cfgstr(cfg);

    if (!handler)
	return -ENXIO;	    /* no handler subtype matches first cfg segment */

    if (!handler->check_config)
	return 0;	    /* OK, no check function */

    /* Advance over handler_name to the handler-specific cfg string */
    cfg = strchrnul(cfg+1, '/');

    err = handler->check_config(cfg, &reason);
    if (err) {
	tcmu_warn("handler %s failed check_config(%s) reason: %s\n",
		    handler->name, cfg, reason?:"none");
	if (reason)
	    sys_mem_free(reason);
    }

    return err;
}

/******** Handlers call these functions ********/

/* Returns zero on success, or -1 on failure */
int
tcmur_register_handler(struct tcmur_handler *handler)
{
	int i;
	int empty_slot = -1;

	for (i = 0; i < (int)ARRAY_SIZE(the_tcmu_handlers); i++) {
		struct tcmur_handler * h = the_tcmu_handlers[i];
		if (!h) {
			if (empty_slot < 0)
				empty_slot = i;
		} else if (!strcmp(h->subtype, handler->subtype)) {
			tcmu_err("Handler %s has already been registered\n",
				 handler->subtype);
			return -1;
		}
	}

	assert_ge(empty_slot, 0);   /* checked before handler_init() */

	tcmu_info("Handler %s registered, slot=%d\n",
		    handler->subtype, empty_slot);
	the_tcmu_handlers[empty_slot] = handler;
	return 0;
}

bool
tcmur_unregister_handler(struct tcmur_handler *handler)
{
	int i;
	for (i = 0; i < (int)ARRAY_SIZE(the_tcmu_handlers); i++) {
		if (the_tcmu_handlers[i] == handler) {
			tcmu_info("Handler %s unregistered, slot=%d\n",
				    the_tcmu_handlers[i]->subtype, i);
			the_tcmu_handlers[i] = NULL;
			return true;
		}
	}
	tcmu_info("Handler %s could not be unregistered, not found\n",
		    handler->subtype);
	return false;
}

extern void *tcmur_dev_get_private(struct tcmu_device *);
void *tcmur_dev_get_private(struct tcmu_device *dev)
{
	return dev->hm_private;
}

extern void tcmur_dev_set_private(struct tcmu_device *, void *);
void tcmur_dev_set_private(struct tcmu_device *dev, void *private)
{
	dev->hm_private = private;
}

void tcmu_dev_set_num_lbas(struct tcmu_device *dev, uint64_t num_lbas)
{
	dev->num_lbas = num_lbas;
}

uint64_t tcmu_dev_get_num_lbas(struct tcmu_device *dev)
{
	return dev->num_lbas;
}

void tcmu_dev_set_block_size(struct tcmu_device *dev, uint32_t block_size)
{
	dev->block_size = block_size;
}

uint32_t tcmu_dev_get_block_size(struct tcmu_device *dev)
{
	return dev->block_size;
}

void tcmu_dev_set_max_xfer_len(struct tcmu_device *dev, uint32_t len)
{
	dev->max_xfer_len = len;    /* in units of block_size */
}

uint32_t tcmu_dev_get_max_xfer_len(struct tcmu_device *dev)
{
	return dev->max_xfer_len;
}

void tcmu_dev_set_write_cache_enabled(struct tcmu_device *dev, bool enabled)
{
	dev->write_cache_enabled = enabled;
}

bool tcmu_dev_get_write_cache_enabled(struct tcmu_device *dev)
{
	return dev->write_cache_enabled;
}

char *tcmu_dev_get_cfgstring(struct tcmu_device *dev)
{
	return dev->cfgstring;
}

void tcmur_cmd_complete(struct tcmu_device *dev, void *data, int sts)
{
    struct tcmur_cmd *tcmur_cmd = data;
    tcmur_cmd->done(dev, tcmur_cmd, sts);
}

/******** Client calls these functions to invoke handlers ********/

/* tcmur_read(), tcmur_write() and tcmur_flush() start I/O operations to the
 * specified minor.  Errors in the I/O start process are reported by a -errno
 * return from these calls.  A return value of zero denotes a successful I/O
 * start, in which case there will be a later completion call to cmd->done(),
 * which may report either an "sts" error, or success (denoted by TCMU_STS_OK).
 */

error_t
tcmur_read(int minor, struct tcmur_cmd * cmd,
	    struct iovec * iov, size_t niov, size_t nbyte, off_t seekpos)
{
    off_t endpos = seekpos + (ssize_t)nbyte;
    struct tcmu_device * dev = device_of_minor(minor);
    error_t err = 0;
    tcmur_status_t sts;

    if (!dev)
	return -ENODEV;	    /* nonexistent minor */
    if (!dev->rhandler->read)
	return -ENXIO;	    /* handler has no read function */
    if (endpos < seekpos)
	return -EINVAL;	    /* I/O exceeding device bounds */
    if (endpos > (off_t)(dev->num_lbas * dev->block_size))
	return -EINVAL;	    /* I/O exceeding device bounds */

    tcmu_io_trace_dev(dev, "READ %lu bytes at offset %lu\n", nbyte, seekpos);

    /* XXX Do handlers look at these, or only at the passed arguments? */
    cmd->iovec = iov;
    cmd->iov_cnt = niov;

    sts = dev->rhandler->read(dev, cmd, iov, niov, nbyte, seekpos);
    if (sts == TCMU_STS_OK) {
	//XXX should let our caller do this
	if (dev->rhandler->nr_threads)
	    tcmur_cmd_complete(dev, cmd, sts);
    } else {
	tcmu_io_trace_dev(dev, "READ ERROR sts=%d", sts);
	if (sts == TCMU_STS_NO_RESOURCE)
	    err = -ENOMEM;
	else
	    err = -EIO;
    }

    return err;
}

error_t
tcmur_write(int minor, struct tcmur_cmd * cmd,
	    struct iovec * iov, size_t niov, size_t nbyte, off_t seekpos)
{
    off_t endpos = seekpos + (ssize_t)nbyte;
    struct tcmu_device * dev = device_of_minor(minor);
    error_t err = 0;
    tcmur_status_t sts;

    if (!dev)
	return -ENODEV;	    /* nonexistent minor */
    if (!dev->rhandler->write)
	return -ENXIO;	    /* handler has no write function */
    if (endpos < seekpos)
	return -EINVAL;	    /* I/O exceeding device bounds */
    if (endpos > (off_t)(dev->num_lbas * dev->block_size))
	return -EINVAL;	    /* I/O exceeding device bounds */

    tcmu_io_trace_dev(dev, "WRITE %lu bytes at offset %lu\n", nbyte, seekpos);

    /* XXX Do handlers look at these, or only at the passed arguments? */
    cmd->iovec = iov;
    cmd->iov_cnt = niov;

    sts = dev->rhandler->write(dev, cmd, iov, niov, nbyte, seekpos);
    if (sts == TCMU_STS_OK) {
	//XXX should let our caller do this
	if (dev->rhandler->nr_threads)
	    tcmur_cmd_complete(dev, cmd, sts);
    } else {
	tcmu_io_trace_dev(dev, "WRITE ERROR sts=%d", sts);
	if (sts == TCMU_STS_NO_RESOURCE)
	    err = -ENOMEM;
	else
	    err = -EIO;
    }

    return err;
}

error_t
tcmur_flush(int minor, struct tcmur_cmd * cmd)
{
    struct tcmu_device * dev = device_of_minor(minor);
    error_t err = 0;
    tcmur_status_t sts;

    if (!dev)
	return -ENODEV;	    /* nonexistent minor */
    if (!dev->rhandler->flush)
	return 0;	    //XXX right?

    tcmu_io_trace_dev(dev, "flush\n");

    sts = dev->rhandler->flush(dev, cmd);
    if (sts == TCMU_STS_OK) {
	//XXX should let our caller do this
	if (dev->rhandler->nr_threads)
	    tcmur_cmd_complete(dev, cmd, sts);
    } else {
	tcmu_io_trace_dev(dev, "FLUSH ERROR sts=%d", sts);
	if (sts == TCMU_STS_NO_RESOURCE)
	    err = -ENOMEM;
	else
	    err = -EIO;
    }

    return err;
}

/******** Client calls these functions to manage devices and handlers ********/

const char *
tcmur_get_dev_name(int minor)
{
    struct tcmu_device * dev = device_of_minor(minor);
    if (!dev)
	return NULL;
    return dev->dev_name;
}

ssize_t
tcmur_get_size(int minor)
{
    struct tcmu_device * dev = device_of_minor(minor);
    if (!dev)
	return -ENODEV;	    /* nonexistent minor */
    return (ssize_t)(dev->num_lbas * dev->block_size);
}

ssize_t
tcmur_get_block_size(int minor)
{
    struct tcmu_device * dev = device_of_minor(minor);
    if (!dev)
	return -ENODEV;	    /* nonexistent minor */
    return dev->block_size;
}

ssize_t
tcmur_get_max_xfer(int minor)
{
    struct tcmu_device * dev = device_of_minor(minor);
    if (!dev)
	return -ENODEV;	    /* nonexistent minor */
    return dev->max_xfer_len;
}

/* Add a block device of the given minor number.
 * cfg starts with subtype followed by the config string for the underlying
 * tcmu-runner handler:
 *			    /subtype/handler_cfg
 */
error_t
tcmur_device_add(int minor, const char * cfg)
{
    struct tcmu_device * dev;
    size_t size;
    error_t err;

    if (minor >= (int)ARRAY_SIZE(the_tcmu_devices))
	return -ENODEV;

    if (device_of_minor(minor))
	return -EBUSY;

    /* Check that cfg refers to a loaded handler */
    err = tcmur_check_config(cfg);
    if (err)
	return err;

    //XXX Do we need to check for a duplicate cfgstring and -EEXIST?
    //    I think the handler has to do it, because only the handler
    //    knows how to tell if two cfgstrings refer to the same device.
    //	  (But we could still check for identical *strings* here)

    dev = sys_mem_zalloc(sizeof(*dev));
    dev->rhandler = handler_of_cfgstr(cfg);
    assert_ne(dev->rhandler, 0);

    /* Advance over handler_name to the handler-specific cfg string */
    cfg = strchrnul(cfg+1, '/');

    snprintf(dev->dev_name, sizeof(dev->dev_name), "%s%03u",
				    dev->rhandler->subtype, minor);
    memset(dev->cfgstring_orig, 0, sizeof(dev->cfgstring_orig));
    strncpy(dev->cfgstring_orig, cfg, sizeof(dev->cfgstring_orig)-1);
    memcpy(dev->cfgstring, dev->cfgstring_orig, sizeof(dev->cfgstring));

    if (dev->rhandler->open) {
	err = dev->rhandler->open(dev, 0);	    /* Call into handler */
	if (err) {
	    tcmu_dev_err(dev, "%s handler->open(%s) returned err=%d\n",
			  dev->rhandler->name, dev->dev_name, err);
	    goto fail_free;
	}
    }

    /* handler->open() might corrupt the config string using strtok() */
    memcpy(dev->cfgstring, dev->cfgstring_orig, sizeof(dev->cfgstring));

#if 0	//XXX no tcmu_cfgfs_dev_get_*()
	block_size = tcmu_cfgfs_dev_get_attr_int(dev, "hw_block_size");
	if (block_size <= 0) {
		tcmu_dev_err(dev, "Could not get hw_block_size\n");
		return -EINVAL;
	}

	dev_size = tcmu_cfgfs_dev_get_info_u64(dev, "Size", &ret);
	if (ret < 0) {
		tcmu_dev_err(dev, "Could not get device size\n");
		return -EINVAL;
	}

	max_sectors = tcmu_cfgfs_dev_get_attr_int(dev, "hw_max_sectors");
	if (max_sectors < 0) {
		tcmu_dev_err(dev, "Bad hw_max_sectors (max_xfer_len) %ld\n",
				    max_sectors);
		return -EINVAL;
	}
#else	//XXX

    if (!tcmu_dev_get_block_size(dev)) {
	sys_notice("Using default block size=%d", 4096);
	tcmu_dev_set_block_size(dev, 4096);		//XXX 4 KiB
    }

    if (!tcmu_dev_get_num_lbas(dev)) {
	sys_notice("Using default nblocks=%d", 262144);
	tcmu_dev_set_num_lbas(dev, 262144);		//XXX 1 GiB
    }

    if (!tcmu_dev_get_max_xfer_len(dev)) {
	sys_notice("Using max I/O size=%d", 1024*1024);
	tcmu_dev_set_max_xfer_len(dev, 1024*1024);	//XXX 1 MiB
    }

#endif

    size = dev->num_lbas * dev->block_size;

    tcmu_info("Handler %s attach target %s size %"PRIu64" block_size %d\n",
	       dev->rhandler->name, dev->dev_name, size, dev->block_size);

    the_tcmu_devices[minor] = dev;

    return 0;

fail_free:
    sys_mem_free(dev);
    return err;
}

error_t
tcmur_device_remove(int minor)
{
    struct tcmu_device * dev = device_of_minor(minor);
    if (!dev)
	return -ENODEV;

    tcmu_info("handler %s destroy tgt: %s\n",
	       dev->rhandler->name, dev->dev_name);

    the_tcmu_devices[minor] = NULL;

    if (dev->rhandler->close)
	dev->rhandler->close(dev);

    sys_mem_free(dev);
    return 0;
}

/* Return 0 if OK, otherwise -errno.
 * The name of a handler file is expected to be:
 *	    concatenate(handler_prefix, subtype, ".so")
 */
error_t
tcmur_handler_load(const char * subtype)
{
    char *error;
    char *path;
    void *handle;
    int (*handler_init)(void);
    int ret;
    int i;
    struct tcmur_handler * h;

    h = tcmur_find_handler(subtype);
    if (h) {
	tcmu_err("%s: Handler %s is already registered\n", subtype, h->subtype);
	return -EEXIST;
    }

    for (i = 0; i < (int)ARRAY_SIZE(the_tcmu_handlers); i++)
	if (!the_tcmu_handlers[i])
	    break;	/* found an empty slot */

    if (i >= (int)ARRAY_SIZE(the_tcmu_handlers)) {
	tcmu_err("Out of handler slots trying to register %s\n", subtype);
	return -ENOSPC;
    }

    ret = sys_asprintf(&path, "%s%s.so", handler_prefix, subtype);
    if (ret < 0)
	return -errno;

    handle = dlopen(path, RTLD_NOW|RTLD_LOCAL);
    if (!handle) {
	tcmu_err("Could not open handler at %s: %s\n", path, dlerror());
	ret = -ENOENT;
	goto out_free;
    }

    dlerror();
    handler_init = dlsym(handle, "handler_init");
    if ((error = dlerror())) {
	tcmu_err("dlsym failure on %s: (%s)\n", path, error);
	ret = -EBADF;
	goto err_close;
    }

    ret = handler_init();
    if (ret) {
	tcmu_err("handler_init failed on path %s\n", path);
	ret = -EIO;
	goto err_close;
    }

out_free:
    sys_mem_free(path);
    return ret;

err_close:
    dlclose(handle);
    goto out_free;
}

error_t
tcmur_handler_unload(const char * subtype)
{
    int slot;
    int i;

    struct tcmur_handler * h = tcmur_find_handler_slot(subtype, &slot);
    if (!h) {
	tcmu_err("Handler %s is not registered\n", subtype);
	return -ENOENT;
    }

    for (i = 0; i < (int)ARRAY_SIZE(the_tcmu_devices); i++)
	if (the_tcmu_devices[i] && the_tcmu_devices[i]->rhandler == h) {
	    tcmu_err("Handler %s has existing devices\n", subtype);
	    return -EBUSY;
	}

    //XXX dlclose()
    tcmur_unregister_handler(h);
    return 0;
}

/* Client can specify callbacks for device add/remove events */
error_t
libtcmur_init(const char * hprefix)
{
    if (hprefix)
	handler_prefix = hprefix;
    return 0;
}

error_t
libtcmur_exit(void)
{
    int i;
    for (i = 0; i < (int)ARRAY_SIZE(the_tcmu_handlers); i++) {
	if (the_tcmu_handlers[i])
	    return -EBUSY;
    }
    return 0;
}
